x/config/config.go

```diff
@@ -68,39 +69,88 @@ func WrapStreamDialer(dialer transport.StreamDialer, transportConfig string) (tr
 	return dialer, nil
 }
 
-func newStreamDialerFromPart(innerDialer transport.StreamDialer, oneDialerConfig string) (transport.StreamDialer, error) {
+type StreamDialer struct {
+	transport.StreamDialer
+	Config string
+}
+
+func (sd *StreamDialer) SanitizedConfig(scheme string, oneDialerConfig string) string {
+	var sanitizedConfig string
+	u, err := parseConfigPart(oneDialerConfig)
+	if err != nil {
+		return ""
+	}
+	switch scheme {
+	case "socks5":
+		if u.User != nil {
+			u.User = url.User("redacted")
+			sanitizedConfig = u.String()
+		} else {
+			sanitizedConfig = oneDialerConfig
+		}
+	case "ss":
+		if u.User != nil {
+			u.User = url.User("redacted")
+			sanitizedConfig = u.String()
+		} else {
+			sanitizedConfig = oneDialerConfig
+		}
+	case "split":
+		sanitizedConfig = oneDialerConfig
+	case "tls":
+		sanitizedConfig = oneDialerConfig
+	case "tlsfrag":
+		sanitizedConfig = oneDialerConfig
+	default:
+		sanitizedConfig = scheme + ":redacted"
+	}
+
+	if sd.Config == "" {
+		return sanitizedConfig
+	} else {
+		return sd.Config + "|" + sanitizedConfig
+	}
+}
+
+func newStreamDialerFromPart(innerDialer *StreamDialer, oneDialerConfig string) (*StreamDialer, error) {
 	url, err := parseConfigPart(oneDialerConfig)
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse config part: %w", err)
 	}
 
 	// Please keep scheme list sorted.
-	switch strings.ToLower(url.Scheme) {
+	scheme := strings.ToLower(url.Scheme)
+	switch scheme {
 	case "socks5":
 		endpoint := transport.StreamDialerEndpoint{Dialer: innerDialer, Address: url.Host}
-		return socks5.NewStreamDialer(&endpoint)
+		dialer, err := socks5.NewStreamDialer(&endpoint)
+		return &StreamDialer{StreamDialer: dialer, Config: innerDialer.SanitizedConfig(scheme, oneDialerConfig)}, err
```

Comment:
This is duplicated code that makes it harder to add new configs. Move to `WrapStreamDialer` and revert `newStreamDialerFromPart`.